beacon-chain/attestation/service.go

```diff
@@ -0,0 +1,110 @@
+// Package attestation defines the life-cycle and status of single and aggregated attestation.
+package attestation
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/ethereum/go-ethereum/event"
+	"github.com/prysmaticlabs/prysm/beacon-chain/types"
+	"github.com/sirupsen/logrus"
+)
+
+var log = logrus.WithField("prefix", "attestation")
+
+// Service represents a service that handles the internal
+// logic of managing aggregated attestation.
+type Service struct {
+	ctx           context.Context
+	cancel        context.CancelFunc
+	handler       *Handler
+	broadcastFeed *event.Feed
+	broadcastChan chan *types.Attestation
+	incomingFeed  *event.Feed
+	incomingChan  chan *types.Attestation
+}
+
+// Config options for the service.
+type Config struct {
+	Handler                 *Handler
+	ReceiveAttestationBuf   int
+	BroadcastAttestationBuf int
+}
+
+// NewService instantiates a new service instance that will
+// be registered into a running beacon node.
+func NewService(ctx context.Context, cfg *Config) *Service {
+	ctx, cancel := context.WithCancel(ctx)
+	return &Service{
+		ctx:           ctx,
+		cancel:        cancel,
+		handler:       cfg.Handler,
+		broadcastFeed: new(event.Feed),
+		broadcastChan: make(chan *types.Attestation, cfg.BroadcastAttestationBuf),
+		incomingFeed:  new(event.Feed),
+		incomingChan:  make(chan *types.Attestation, cfg.ReceiveAttestationBuf),
+	}
+}
+
+// Start an attestation service's main event loop.
+func (a *Service) Start() {
+	log.Info("Starting service")
+	go a.aggregateAttestations()
+}
+
+// Stop the Attestation service's main event loop and associated goroutines.
+func (a *Service) Stop() error {
+	defer a.cancel()
+	log.Info("Stopping service")
+	return nil
+}
+
+// IncomingAttestationFeed returns a feed that any service can send incoming p2p attestations into.
+// The attestation service will subscribe to this feed in order to relay incoming attestations.
+func (a *Service) IncomingAttestationFeed() *event.Feed {
+	return a.incomingFeed
+}
+
+// ContainsAttestation checks if an attestation has already been received and aggregated.
+func (a *Service) ContainsAttestation(bitfield []byte, h [32]byte) (bool, error) {
+	attestation, err := a.handler.getAttestation(h)
+	if err != nil {
+		return false, fmt.Errorf("could not get attestation from DB: %v", err)
+	}
+	savedAttestationBitfield := attestation.AttesterBitfield()
+	for i := 0; i < len(bitfield); i++ {
+		if bitfield[i]&savedAttestationBitfield[i] != 0 {
```

Comment:
Is this accurate ? You could have one byte of the bitfield  that is the same as the saved `attesterBitfield` but the next byte could be different from each other. According to the logic here it will return true when it should be false. It should instead be 
```golang
for i := 0; i < len(bitfield); i++ {
		if bitfield[i]&savedAttestationBitfield[i] == 0 {
			return false, nil
		}
	}
	return true, nil
```

Reply:
We just need one byte to be the same to to know if this attestation has been aggregated before.

bitfield will always be a bitfield with single bit flipped. 

For example:
1.) Return true on first byte, skip aggregating this attestation, we have done it already
incoming bitfield: 00000001 00000000 00000000 00000000
aggregated bitfield: 00010001 11110000 00011000 000111100

1.) Return false, aggregate this attestation, we haven't seen it before
incoming bitfield: 00000010 00000000 00000000 00000000
aggregated bitfield: 00010001 11110000 00011000 000111100

Reply:
ah , k that clears it up . For some reason I thought we were comparing something like this 
`00010001 11110000 00011000 000111100` and
`00010001 11110000 00011000 000111000`, in that case your method is fine

Reply:
> bitfield will always be a bitfield with single bit flipped.

If so, can I change the function to accept a single integer representing the validator's index, and check that bit directly?

Reply:
Ah, I see. The attestation conforms to `AggregatedAttestation`. Isn't this a large waste of space?

Saythere's 100k validators, each bitfield takes up 100kilobytes of space, when it could be represented by a single integer?

Reply:
Sounds good. This may be a medium sized change, so I'll open a ticket for now.